### DISCUSSION TITLE
Installing spacy-transformers on mac with m1 chip fails

### DISCUSSION THREAD

**Discussion by JeffryCA** [explosion/spaCy#7835] [Help: Installation] [answered]
## How to reproduce the behaviour
pip install spacy-transformers

and also failing (no error it just gets stuck every time):

python -m spacy download de_dep_news_trf
 

## Your Environment

* Operating System: OS 11.2.3
* Python Version Used: 3.9 (conda)
* spaCy Version Used: 3.0.5

**Comment by adrianeboyd** [comment, marked as answer]
Hi, it looks like the current `spacy-transformers` (v1.0.1) was released before there was a version of `tokenizers` that can use a version of rust with support for the M1 architecture, so there's not a version of `tokenizers`+`transformers` that can be installed according to the `spacy-transformers` v1.0.1 requirements.

If you want to use it in the meanwhile, you can install from the current `master` branch in the github repo.

First, you will need to install rust (for `tokenizers` and `spacy-alignments`) and it's best to install pytorch separately first (probably: `conda install pytorch -c conda-forge`).

Then I think `pip install https://github.com/explosion/spacy-transformers/archive/master.zip` will work from the current `master` branch. There aren't many changes beyond extending the supported `transformers` version.

We can hopefully release the next version of `spacy-transformers` soon (will be v1.0.2) with support for newer versions of `tokenizers`.

**Comment by KTRosenberg** [comment]
Has anything changed with regard to this issue? I tried installing 
`python3 -m spacy download en_core_web_trf` in a condo environment, but get a large failure message saying that rust's environment variables need to be set:

```
  ERROR: Command errored out with exit status 1:
   command: /Users/username/miniforge3/envs/TEST/bin/python /Users/username/miniforge3/envs/TEST/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /var/folders/c6/gmbh3q_56nd078hcmng9z64h0000gn/T/tmpsi64l5j9
       cwd: /private/var/folders/c6/gmbh3q_56nd078hcmng9z64h0000gn/T/pip-install-7hurf5tb/spacy-alignments_edc726e2c8da4eef86fbfad8191f38af
  Complete output (21 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.macosx-11.0-arm64-3.9
  creating build/lib.macosx-11.0-arm64-3.9/spacy_alignments
  copying spacy_alignments/__init__.py -> build/lib.macosx-11.0-arm64-3.9/spacy_alignments
  creating build/lib.macosx-11.0-arm64-3.9/spacy_alignments/tests
  copying spacy_alignments/tests/__init__.py -> build/lib.macosx-11.0-arm64-3.9/spacy_alignments/tests
  copying spacy_alignments/tests/test_main.py -> build/lib.macosx-11.0-arm64-3.9/spacy_alignments/tests
  running egg_info
  writing spacy_alignments.egg-info/PKG-INFO
  writing dependency_links to spacy_alignments.egg-info/dependency_links.txt
  writing top-level names to spacy_alignments.egg-info/top_level.txt
  reading manifest file 'spacy_alignments.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  writing manifest file 'spacy_alignments.egg-info/SOURCES.txt'
  running build_ext
  running build_rust
  error: Don't know the correct rust target for system type arm64-apple-darwin20.0.0. Please set the CARGO_BUILD_TARGET environment variable.
  ----------------------------------------
  ERROR: Failed building wheel for spacy-alignments
  ```
  
  The same goes for spacy-tokenizers.

Is there any way to resolve this?

**Reply by adrianeboyd** [reply]
Hi, I don't think this is the same error as the original post. With just `pip` this works fine, so I think this is something related to the conda env, I'm not exactly sure what, but it's confused between `arm64` (from python) vs. `aarch64` (from rust). Try setting this first:

```none
export CARGO_BUILD_TARGET="aarch64-apple-darwin"
```

We just added `spacy-alignments` and `spacy-transformers` to `conda-forge` but it didn't automatically add non-`x86_64` builds and I'm not sure whether there's going to be an automatic update that adds them sometime soon or there's something related to the dependencies that's preventing it.

**Reply by KTRosenberg** [reply]
Ah so I should’ve started a new discussion.

So should all the transformers models work at this point? I think my Python environment has been messed-up somehow, as spacy[apple] also doesn’t work using regular pip. (The [apple] part is unrecognized). I am using 3.9.
At some point I attempted with both arm python and x64 python to try and force emulation mode because I am also trying to get a library called “coreferee” working that depends on Spacy 3.1 and tensorflow, and there are additional problems there in arm mode. Then I got illegal hardware instructions.

**Reply by adrianeboyd** [reply]
All the transformers models should work as far as I know. Several spacy developers work full-time on M1s now, and I'm pretty sure that all the spacy libraries work fine in general. I know less about tensorflow because we don't use it in any of the main packages. Most of don't use conda on a daily basis, so sometimes there are separate installation issues which we haven't run into ourselves yet.

Sorry, in my response I meant that this is with `pip install spacy-alignments` in a non-conda venv-only environment vs. `pip install spacy-alignments` in a conda env. You're still using python from conda according to the paths above and I was able to replicate this in a similar brand-new conda env. (`conda install spacy-alignments` won't currently work on an M1 at all because there's no package that meets the requirements.)

I would generally recommend using an architecture-specific python binary so that you can't easily end up with incompatible packages in your environment, and the miniforge3 osx arm64 installer is a good starting point, since there are packages for most spacy libraries already available and also `transformers`, plus now I see that there's a `tensorflow` package. If you're using conda, it's usually best to install as much as possible with conda (`spacy`, `pytorch`, `tensorflow`, `transformers`) and then use pip for the remaining packages as necessary. You can also install `rust` through conda-forge for `spacy-alignments`. And you can always install `thinc-apple-ops` directly with pip with `pip install thinc-apple-ops`.

In general, I suspect that starting over in a new conda env with an arm64-only python binary will help.

**Reply by KTRosenberg** [reply]
I’ll give this another try then.
Just to clarify, do I need to reinstall the rust compiler and install the dependencies for the transformer model separately?

Won’t installing the transformer model still ignore existing installations and fail as shown above? If needed, how do I specify the rust environment variable?

My apoloies if this is too much handholding, but could you show some specific commands I should try? I think another issue I encountered was that conda’s version of pip was ignored. There’s some specific path I am supposed to use.

Also, should this all still work in Python 3.9 and the previous point version of spacy (3.1.4?) on macos11? I will also have to figure out how to overwrite the other 3.9 that I already have.

I’ll come back first after I’ve made an attempt again.

For the record, this is the library causing me trouble:

https://github.com/msg-systems/coreferee

I have been looking forward to a Spacy builtin coreference resolution model to replace this for a while.

**Reply by KTRosenberg** [reply]
Now I get "ModuleNotFoundError: No module named 'keras'" when initializing a model. Strange.

**Reply by KTRosenberg** [reply]
I tried again with just spacy, and now it sort of works, but using the trf model yields the warning,
"  warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling')"
I guess this is okay, but I'm not sure where that came from.

**Reply by KTRosenberg** [reply]
Good news: I got it working:
```
conda install rust 
export CARGO_BUILD_TARGET="aarch64-apple-darwin"
conda install -c conda-forge spacy=3.1.4
```
I hope that build target step becomes unnecessary eventually.

**Reply by jamesonquinn** [reply]
Thanks, @KTRosenberg. I'm still getting the CUDA problem even after installing as you say:

...lib/python3.9/site-packages/torch/autocast_mode.py:141: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
warnings.warn('User provided device_type of 'cuda', but CUDA is not available. Disabling')

Any other clues from your experience would be appreciated.

**Reply by KTRosenberg** [reply]
@jamesonquinn Does that warning actually do anything? As in, does it revert to software processing instead of using the GPU? Spacy actually works if you ignore the warning, but if the GPU doesn't work, that's not great, and I'd like to know how to fix this too.

**Reply by adrianeboyd** [reply]
For reference, this problem first occurs with `setuptools-rust==1.0.0`, which was released about a week ago. Earlier versions of `setuptools-rust` did work as expected, which is why we hadn't seen this problem before. I submitted a bug report and my guess is that it's not going to be difficult to fix (although I'm not familiar enough with rust to know for sure), but if there's no an easy fix, we can consider restricting the version of `setuptools-rust` in the package setup instead.

**Reply by KTRosenberg** [reply]
@adrianeboyd Is that CUDA warning related? Is it something that can be ignored, or does it disable any use of the GPU, in which case, what else do I need to do?

**Reply by jamesonquinn** [reply]
@KTRosenberg I managed to get the warning to go away by reinstalling conda from miniforge. I could tell I needed to do that because activity monitor was showing that my python threads were running under intel emulation. Once I'd done that (and finished re-adding all the deps) it does in fact run faster (about 5x speedup on my task).

**Reply by KTRosenberg** [reply]
@jamesonquinn So you restarted? What is your full installation procedure? (If you could provide a script, that would be great.) I have to use spacy 3.1.4. Is 3.2 required? Do I just start the minigorge installation script?

Do you also set the rust arm flag?

I guess another problem could be the need for some tensorflow installation for another library, which might introduce some doengrades.

**Reply by adrianeboyd** [reply]
There's now a new version of `setuptools-rust` that should fix this particular issue. If you're having other problems, it would be better to open a new discussion thread.
